### PR TITLE
Composite traintuples

### PR DESCRIPTION
**Pull request by AurelienGasser**
**See also**

- [substra-backend PR](https://github.com/SubstraFoundation/substra-backend/pull/13)
- [chaincode "aggregate" PR](https://github.com/SubstraFoundation/substra-chaincode/pull/26)

---

**Tasks**

- [x] Creation of a new structure used for “inputting” a new compositetraintuple
- [x] Creation of a new structure corresponding to elements stored in the ledger for a given compositetraintuple
- [x] Creation of a new structure used for “outputting” a compositetraintuple
- [x] Creation of a smart contract to create an compositetraintuple
- [x] Creation of smart contracts to query an compositetraintuple and query all compositetraintuple
- [x] Create a smart contract to add/get/list compositeAlgo, which have the same structure as a normal Algo (and the same input and output)
- [x] Modification of createTesttuple to make it possible to consider as input a compositetraintuple (on top of a traintuple)
- [x]  Make compositeAlgo the algo type used in compositeTraintuple
- [x] Update the various indexes depending on what we want traintuple/compositeTraintuple queries to return
- [x] Creation of smart contracts similar to logStartTrain, logFailTrain, logSuccesTrain for compositetraintuples
- [x] Add to outputTesttuple a field that represents the traintuple type
- [x] Add composite traintuple to query model smart contracts
- [x] Add testtuples tests (and among those, add one that test status propagation to testtuples)
- [x] Update the list of smartcontracts
- [ ] ~~Include composite traintuple to the computePlan workflow (create and query)~~ TODO In another PR

### CODE REVIEW

**Comment by AurelienGasser** [comment]
Sorry for the noise. I wanted to create a "draft" PR and created a regular PR instead. There's [currently no way](https://github.community/t5/How-to-use-Git-and-GitHub/Feature-Request-Switch-from-ready-to-draft-in-pull-requests/td-p/19107) to revert from a normal PR to a draft PR, so I added the `do not merge` and `work in progress` labels instead.

File: chaincode/input.go
Comment (AurelienGasser) [inline chaincode/input.go]: On second thought, I think `inputModelComposite` would be better, what do you think @thibaultrobert ? IMO having the suffix `Composite` makes it clearer that we're talking about a `Composite`-specific case of an `InModel`.

Comment (AurelienGasser) [reply chaincode/input.go]: I'm gonna go with `Composite` suffix for the "out model" struct for now, I can update depending your response :)

**Comment by AurelienGasser** [comment]
@thibaultrobert Please note that I bumped the `validator` version from v9.29.0 to v9.30.0, because the previous version had a bug (`required_with` would create a panic!)

File: chaincode/input.go
Comment (inelgnu) [inline chaincode/input.go]: seems that you inverted model keys (trunk/head)

Comment (AurelienGasser) [reply chaincode/input.go]: @inalgnu if you're talking about the `required_with` field, it's intentional! It's to make sure you can't pass a `head` model without `trunk` model, and vice versa.

Comment (inelgnu) [reply chaincode/input.go]: my bad I thought it was the json field name.

File: chaincode/output.go
Comment (samlesu) [inline chaincode/output.go]: Should be `compositetraintuple`?
This is currently what is expected by the backend.

File: chaincode/utils.go
Comment (thibaultrobert) [inline chaincode/utils.go]: Not sure about the case. I think it should start with a lower case at least: `dataManager`, `traintuple`, `compositeTraintuple`,....

File: chaincode/input.go
Comment (samlesu) [inline chaincode/input.go]: Should be `OutTrunkModelPermission`?

File: chaincode/utils.go
Comment (AurelienGasser) [reply chaincode/utils.go]: Will do! Is there a particular reason to change the strings to camel case ?

File: chaincode/testtuple.go
Comment (AurelienGasser) [inline chaincode/testtuple.go]: The permissions are checked against the `head` out-model, but the value assigned to the`testtuple.Model` is the `trunk` out-model. Is that intentional ?

File: chaincode/input.go
Comment (AurelienGasser) [reply chaincode/input.go]: That makes sense to me. @thibaultrobert, what do you think?

File: chaincode/output.go
Comment (AurelienGasser) [reply chaincode/output.go]: fixed https://github.com/SubstraFoundation/substra-chaincode/pull/5/commits/e8148ed4fab510f1bf23bfa1a3e58eb95577f5e2

File: chaincode/utils.go
Comment (thibaultrobert) [reply chaincode/utils.go]: After a quick discussion, we'll keep this version.

File: chaincode/input.go
Comment (thibaultrobert) [reply chaincode/input.go]: If it's OK with `substra-backend`, I agree.

File: chaincode/output.go
Comment (AurelienGasser) [reply chaincode/output.go]: I had forgotten the json field :man_facepalming:
Fixed: https://github.com/SubstraFoundation/substra-chaincode/pull/5/commits/51cbefe9e00b224681b1725c4cf85fb013201937

File: chaincode/input.go
Comment (AurelienGasser) [reply chaincode/input.go]: Fixed https://github.com/SubstraFoundation/substra-chaincode/pull/5/commits/521466aa5ec9bd1ac217c82847b35349bb36d273

File: chaincode/go.mod
Comment (thibaultrobert) [inline chaincode/go.mod:9]: Beware, we should update that before merging

File: chaincode/utils.go
Comment (thibaultrobert) [inline chaincode/utils.go:130]: I'm not a fan to use a string to rune to string conversion but I don't have strong arguments to sustain such statement. But just to discuss it, what do you think of something like:
```go
t := strings.SplitN(s,"",2)
return strings.ToLower(t[0]) + t[1]
```

Comment (AurelienGasser) [reply chaincode/utils.go:130]: +1 that makes sense!

Comment (thibaultrobert) [reply chaincode/utils.go:130]: Not that much now that I think about it.... It will panic for a one character string!

Comment (thibaultrobert) [reply chaincode/utils.go:130]:
```go
t := strings.SplitN(s,"",2)
t[0] = strings.ToLower(t[0])
return strings.Join(t, "")
```
Maybe?

**Comment by AurelienGasser** [comment]
This is ready for review / testing :tada: 

**Edit**: There was a bug that [wasn't caught by the unit tests](https://github.com/SubstraFoundation/substra-chaincode/pull/38). It's now fixed.

**Comment by AurelienGasser** [comment]
I ran the end-to-end tests in the following setup, it's :+1: 

- substra: `dev` 
- substra-backend: `dev`
- substra-chaincode `feature/generic-composite`
- substra-tests: `master`

The only tests that fail also fail when I use the `dev` chaincode branch (the failing tests are all related to compute plans) 

It looks safe to merge.

**Review by inelgnu — DISMISSED** [review]

**Comment by AurelienGasser** [comment]
Update: both this PR and the [aggregate PR](https://github.com/SubstraFoundation/substra-chaincode/pull/26) are ready for review / merge. 

Things that are **not** done yet (to do in future PRs?):

- aggregate traintuple permissions (currently disregarded)
- compute plans

**Comment by inelgnu** [comment]
@AurelienGasser it's fine to add permissions and compute plan in another PR. I'm totally Ok with that approach.

File: chaincode/testtuple.go
Comment (inelgnu) [inline chaincode/testtuple.go]: I think the merge is not necessary in this case. 
OutHeadModel is forced to one node, so the intersection with one node will always generate the same node.
I think this is why it was forced to compositeTraintuple.OutHeadModel.Permissions

Comment (AurelienGasser) [reply chaincode/testtuple.go]: Fixed https://github.com/SubstraFoundation/substra-chaincode/pull/5/commits/45748822a3926f04b24151bfd0d66dd8144ea9d1

**Comment by AurelienGasser** [comment]
Update: new PR for aggregate permissions: https://github.com/SubstraFoundation/substra-chaincode/pull/39

File: chaincode/output.go
Comment (inelgnu) [inline chaincode/output.go]: 💯

File: README.md
Comment (thibaultrobert) [inline README.md]: No `Aggregatetuple` smartcontracts listed here.

File: chaincode/algo_aggregate.go
Comment (thibaultrobert) [inline chaincode/algo_aggregate.go:52]: It's no longer the case for any type of algo. When we add the algo in the ledger [line 61](https://github.com/SubstraFoundation/substra-chaincode/blob/5dc099c59b2cb5c1a147f2b134f07f4fae7b21ce/chaincode/algo_aggregate.go#L66), we return directly if there is an error and there is one if the key already exists.
An issue should be opened to discuss if you should either remove this line from the `resgisterXXXAlgo` functions or if we should call `db.Put` rather than `db.Add`.

File: chaincode/algo_composite.go
Comment (thibaultrobert) [inline chaincode/algo_composite.go:52]: See previous comment on the `registerXXXAlgo` functions.

File: chaincode/ledger_db.go
Comment (thibaultrobert) [inline chaincode/ledger_db.go]: I would have prefer a error check then the switch:
```go
if err != nil {
  continue
}
switch modelType {...
```
I find it more explicit that way.

File: chaincode/ledger_db.go
Comment (thibaultrobert) [inline chaincode/ledger_db.go]: Is this type conversion necessary?

File: chaincode/ledger_db.go
Comment (thibaultrobert) [inline chaincode/ledger_db.go]: Maybe use `%v` format verb so you don't have to format `allowedStr` by hand:
```suggestion
	return nil, errors.NotFound(
	"GetOutModelHashDress: Could not find traintuple %s with key \"%s\". Allowed types: %v.",
	modelType,
	traintupleKey,
	allowedAssetTypes)
```

File: chaincode/ledger_db_test.go
Comment (thibaultrobert) [inline chaincode/ledger_db_test.go]: If you follow my suggestion on string formatting, those asserts will have to be updated.

File: chaincode/main_test.go
Comment (thibaultrobert) [inline chaincode/main_test.go:627]: It would be better to seed the rand first. Maybe in the `TestMain` function.

File: chaincode/model_composition_test.go
Comment (thibaultrobert) [inline chaincode/model_composition_test.go]:
```go
case "successParent1", "successParent2", "successBoth":
```

File: chaincode/main_test.go
Comment (thibaultrobert) [reply chaincode/main_test.go:627]: Unnecessary since `math/rand` is by default seeded at 1 (See [here](https://golang.org/pkg/math/rand/#Seed))

**Review by thibaultrobert — APPROVED** [review]

**Pull request merged**
